pcdsdevices/ccm.py

```diff
@@ -773,6 +773,178 @@ def inverse(self, real_pos: namedtuple) -> namedtuple:
         return self.PseudoPosition(energy=energy)
 
 
+class CCMEnergyWithSelfSeed(CCMEnergy):
+    """
+    CCM energy motor with the self-seeding scheme.
+
+    Moves the alio to the requested energy, while also writing to the
+    BeamEnergyRequest PV so that ACR can set the self-seed energy accordingly.
+
+    The goal is to optimize the CCM energy around the self-seed energy. The
+    process is as follow:
+        1. Pass a setpoint energy to both the CCM and self-seeding.
+        2. Wait for self-seeding to stabilize.
+        3. Optimize the CCM throughput at the new energy.
+    There can be several implementations of the optimization step. Look at
+    the specific implementation methods for detailed information.
+
+    Parameters
+    ----------
+    prefix : str
+        The PV prefix of the Alio motor, e.g. XPP:MON:MPZ:07A
+    ipms_prefix : str
+        The EventBuilder PV containing the IPMs information
+    ipm_idx : dict
+        Indices in the EventBuilder array PV that contains the upstream
+        and downstream IPM. Format: {'upstream': 0, 'downstream': 1}
+    hutch : str, optional
+        The hutch we're in. This informs us as to which vernier
+        PVs to write to. If omitted, we can guess this from the
+        prefix.
+    """
+    self_seed = FCpt(BeamEnergyRequest, '{hutch}', kind='normal',
+                     doc='Requests ACR to move the self-seed energy.')
+
+    # IPMs intensities from the EventBuilder IOC. We need beam-synchronous IPM
+    # information, hence the use of the EventBuilder PV.
+    # See https://confluence.slac.stanford.edu/display/PCDS/Notes+on+the+Timetool+IOC
+    ipms = UCpt(EpicsSignalRO)
+
+    # These are duplicate warnings with main energy motor
+    _enable_warn_constants: bool = False
+    hutch: str
+
+    tab_component_names = True
+
+    def __init__(
+        self,
+        prefix: str,
+        ipms_prefix: str = None,
+        ipm_idx: dict = {'upstream': 0, 'downstream': 1},
```

Comment:
Interesting! I don't think it would have dramatic effect here, but it is definitely worth sticking to good practices! Thanks for calling me out on something I do all the time :-D